gap/attr.gi

```diff
@@ -290,6 +290,177 @@ function(D)
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+  # Now check if the graph is three colourable. This is done by searching
+  # for a maximal independent set where the subgraph induced by removing the
+  # vertices in the set is bipartite.
+  # Need to make a copy in case we are given a mutable digraph
+  D := DigraphImmutableCopy(D);
+  for I in DigraphMaximalIndependentSets(D) do
+    # Check if removing these vertices gives you a bipartite digraph
+    if IsBipartiteDigraph(DigraphRemoveVertices(D, I)) then
+      return 3;
+    fi;
+  od;
+  # This graph is greater than 3 colourable.
+  return infinity;
+end
+);
+
+InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
+[IsDigraph, IsDigraphColouringAlgorithm and IsDigraphColouringAlgorithmByskov],
+function(D, Byskov)
+  local n, a, vertices, x, s, i, j, I, s_copy, subset_iter, induced_subgraph,
+  index_subsets, vertex_copy;
+
+  n := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif n = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+  vertices := DigraphVertices(D);
+  # Store current best colouring for each subset
+  x := [1 .. 2 ^ n];
+  # Empty set is 0 colourable
+  x[1] := 0;
+  # Function to index the subsets of the vertices of D
+  index_subsets := set -> Sum(set, x -> 2 ^ (x - 1)) + 1;
+  subset_iter := IteratorOfCombinations(vertices);
+  # Skip the first one, which should be the empty set
+  NextIterator(subset_iter);
+  # First find the 3 colourable subgraphs of D
+  for s in subset_iter do
+    i := index_subsets(s);
+    x[i] := infinity;
+    a := DIGRAPHS_UnderThreeColourable(InducedSubdigraph(D, s));
+    # Mark this as three colourable if it is.
+    if a < x[i] then
+      x[i] := a;
+    fi;
+  od;
+  # Process 4 colourable subgraphs
+  for I in DigraphMaximalIndependentSets(D) do
+    vertex_copy := ShallowCopy(vertices);
+    SubtractSet(vertex_copy, I);
+    # Iterate over all subsets of V(D) \ I
+    for s in IteratorOfCombinations(vertex_copy) do
+      i := index_subsets(s);
+      if x[i] = 3 then
+        s_copy := ShallowCopy(s);
+        UniteSet(s_copy, I);
+        j := index_subsets(s_copy);
+        if x[j] > 4 then
+          x[j] := 4;
+        fi;
+      fi;
+    od;
+  od;
+  # Iterate over all vertex subsets.
+  subset_iter := IteratorOfCombinations(vertices);
+  # Skip the first one, which should be the empty set.
+  NextIterator(subset_iter);
```

Comment:
Again, an `IsEmpty` check would be good here.

Reply:
...unless there's anything in the doc that promises that this is the empty set.

Reply:
Done!